proof/refine/AARCH64/ArchKHeap_R.thy

```diff
@@ -0,0 +1,413 @@
+  shows "\<lbrace>valid_global_refs'\<rbrace> f \<lbrace>\<lambda>_. valid_global_refs'\<rbrace>"
+  apply (simp add: valid_global_refs'_def valid_refs'_def global_refs'_def valid_cap_sizes'_def)
+  apply (rule hoare_lift_Pf [where f="ksArchState"])
+   apply (rule hoare_lift_Pf [where f="ksIdleThread"])
+    apply (rule hoare_lift_Pf [where f="irq_node'"])
+     apply (rule hoare_lift_Pf [where f="gsMaxObjectSize"])
+      apply (wp ctes hoare_vcg_const_Ball_lift arch idle irqn maxObj)+
+  done
+
+lemma valid_arch_state_lift':
+  assumes typs: "\<And>T p P. f \<lbrace>\<lambda>s. P (typ_at' T p s)\<rbrace>"
+  assumes arch: "\<And>P. f \<lbrace>\<lambda>s. P (ksArchState s)\<rbrace>"
+  assumes vcpu: "\<And>P p. f \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  shows "f \<lbrace>valid_arch_state'\<rbrace>"
+  apply (simp add: valid_arch_state'_def valid_asid_table'_def vspace_table_at'_defs)
+  apply (wp_pre, wps arch)
+   apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_const_imp_lift vcpu[unfolded pred_conj_def]
+                 split: option.split)
+  apply (clarsimp simp: pred_conj_def)
+  done
+
+lemma set_ntfn_hyp[wp]:
+  "setNotification ptr val \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  unfolding setNotification_def
+  by (wpsimp wp: setObject_ko_wp_at simp: gen_objBits_simps, rule refl, simp)
+     (clarsimp simp: is_vcpu'_def ko_wp_at'_def obj_at'_def)
+
+lemma set_ep_hyp[wp]:
+  "setEndpoint ptr val \<lbrace>\<lambda>s. P (ko_wp_at' (is_vcpu' and hyp_live') p s)\<rbrace>"
+  unfolding setEndpoint_def
+  by (wpsimp wp: setObject_ko_wp_at simp: objBits_simps', rule refl, simp)
+     (clarsimp simp: is_vcpu'_def ko_wp_at'_def obj_at'_def)
+
+lemma idle_is_global[KHeap_R_assms, intro!]:
+  "ksIdleThread s \<in> global_refs' s"
+  by (simp add: global_refs'_def)
+
+end
+
+interpretation KHeap_R?: KHeap_R
+proof goal_cases
+  interpret Arch .
+  case 1 show ?case by (intro_locales; (unfold_locales; fact KHeap_R_assms)?)
+qed
+
+context Arch begin arch_global_naming
+
+named_theorems KHeap_R_assms_2
+
+lemmas setEndpoint_valid_globals[KHeap_R_assms_2, wp]
+  = valid_global_refs_lift'[OF set_ep_ctes_of set_ep_arch'
+                               setEndpoint_it setEndpoint_ksInterruptState]
+
+lemma set_ntfn_global_refs'[KHeap_R_assms_2, wp]:
+  "\<lbrace>valid_global_refs'\<rbrace> setNotification ptr val \<lbrace>\<lambda>_. valid_global_refs'\<rbrace>"
+  by (rule valid_global_refs_lift'; wp)
+
+crunch setEndpoint, setNotification
+  for valid_arch'[wp]: valid_arch_state'
+  (wp: valid_arch_state_lift')
+
+lemmas [KHeap_R_assms_2] = setEndpoint_valid_arch' setNotification_valid_arch'
+
+lemmas setObject_typ_ats[wp] = typ_at_lifts[OF setObject_typ_at']
+
+lemmas doMachineOp_typ_ats[wp] = typ_at_lifts[OF doMachineOp_typ_at']
+
+lemmas setEndpoint_typ_ats[wp] = typ_at_lifts[OF setEndpoint_typ_at']
```

Comment:
I just had the thought that a lot of what `typ_at_lifts` is doing is generic. Do we want to split it out so those parts can be done in generic files or would that be making it more complicated for not much gain?

Reply:
Sounds good and the consistent naming scheme will definitely be a help.

Just a note for the future while this is on my mind, `typ_at_lifts` might be one of the pain points for when it's time to arch-split MCS. We made some pretty significant changes there that ended up being a big improvement but which will mean that what's done on master to it won't map across nicely, if at all. I'm not sure what the best path forwards for this specifically will be, it's possible that backporting the `typ_at_lifts` changes to `master` will help but that also might be more work than it's worth.

Reply:
My knowledge of MCS is very poor, not sure what we can do about it. How different is it over there?

Reply:
Very different, see https://github.com/seL4/l4v/blob/rt/proof/refine/ARM/Invariants_H.thy#L3035. There's a couple of locales that are built up to contain a larger collection of lifting lemmas than are in `typ_at_lifts` in master, and that are then interpreted as needed instead of being instantiated with `OF`.

Some of this is more complicated than it would be on master due to MCS specifics around `sc_at_n`, but I guess at the very least there will need to be an arch version of the locales split out.